storage-node/packages/colossus/lib/sync.js

```diff
@@ -19,73 +19,100 @@
 'use strict'
 
 const debug = require('debug')('joystream:sync')
+const _ = require('lodash')
 
-async function syncCallback(api, storage) {
-  // The first step is to gather all data objects from chain.
-  // TODO: in future, limit to a configured tranche
-  // FIXME this isn't actually on chain yet, so we'll fake it.
-  const knownContentIds = (await api.assets.getKnownContentIds()) || []
+const MAX_CONCURRENT_SYNC_ITEMS = 15
+const contentBeingSynced = new Map()
 
+async function syncCallback(api, storage) {
+  const knownContentIds = await api.assets.getKnownContentIds()
   const roleAddress = api.identities.key.address
   const providerId = api.storageProviderId
 
-  // Iterate over all sync objects, and ensure they're synced.
-  const allChecks = knownContentIds.map(async (contentId) => {
-    // eslint-disable-next-line prefer-const
-    let { relationship, relationshipId } = await api.assets.getStorageRelationshipAndId(providerId, contentId)
-
-    // get the data object
-    // make sure the data object was Accepted by the liaison,
-    // don't just blindly attempt to fetch them
-
-    let fileLocal
-    try {
-      // check if we have content or not
-      const stats = await storage.stat(contentId)
-      fileLocal = stats.local
-    } catch (err) {
-      // on error stating or timeout
-      debug(err.message)
-      // we don't have content if we can't stat it
-      fileLocal = false
-    }
-
-    if (!fileLocal) {
+  // Iterate over all objects, and start syncing if required.
+  // compile list of already syncedIds (as reported by storage
+  // subsytem). The only async part here is resolving content id
+  // by storage to ipfs cid, maybe we can resolve them locally
+  // and cache result to simplify async code below and reduce
+  // queries
+
+  // Since we are limiting concurrent content ids being synced, to ensure
+  // better distribution of content across storage nodes during a potentially long
+  // sync process we don't want all nodes to replicate items in the same order, so
+  // we simply shuffle ids around.
+  const shuffledIds = _.shuffle(knownContentIds)
+
+  const syncedIds = (
+    await Promise.all(
+      shuffledIds.map(async (contentId) => {
+        // TODO: get the data object
+        // make sure the data object was Accepted by the liaison,
+        // don't just blindly attempt to fetch them
+
+        try {
+          const { synced, syncing } = await storage.syncStatus(contentId)
+
+          if (synced) {
+            return contentId
+          }
+
```

Comment:
Do we need to check 'synced' again here?

Reply:
Improved the if/else blocks in
0a6d836